pennylane/data/attributes/array.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains DatasetAttribute definition for numpy arrays."""
+
+from typing import Optional
+import numpy
+from numpy._typing._array_like import _SupportsArray
+from numpy._typing._nested_sequence import _NestedSequence
+from numpy.typing import ArrayLike
+
+from pennylane.data.base.attribute import DatasetAttribute, AttributeInfo
+from pennylane.data.base.hdf5 import HDF5Array, HDF5Group
+import pennylane.math
+
+
+class DatasetArray(DatasetAttribute[HDF5Array, numpy.ndarray, ArrayLike]):
+    """
+    Attribute type for objects that implement the Array protocol, including numpy arrays.
+    """
+
+    type_id = "array"
+
+    def __post_init__(self, value: ArrayLike, info: AttributeInfo | None) -> None:
+        super().__post_init__(value, info)
+
+        self.info["array_interface"] = pennylane.math.get_interface(value)
+        if hasattr(value, "requires_grad"):
+            self.info["requires_grad"] = value.requires_grad
+
+    def hdf5_to_value(self, bind: HDF5Array) -> numpy.ndarray:
```

Comment:
@timmysilv, we now support loading array using the different interfaces

Reply:
nice! simple and gets the job done. I'm a bit worried that we're half-generalizing here, so maybe we should reduce the supposed support. Jax and TF are their own beasts, so let's ignore them for now, but (conveniently!) Torch also uses the `requires_grad` attribute. Unfortunately, it does _not_ play nice:
```pycon
>>> x = torch.tensor(12.3, requires_grad=True)
>>> arr = qml.data.attribute(x)
RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.
```

<details>
<summary>rest of stack trace if you're curious</summary>

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/base/attribute.py", line 412, in attribute
    return match_obj_type(val)(
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/base/attribute.py", line 269, in __init__
    self._bind = self._set_value(value, info, parent, key)
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/base/attribute.py", line 331, in _set_value
    new_bind = self.value_to_hdf5(parent, key, value)
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/attributes/array.py", line 53, in value_to_hdf5
    bind_parent[key] = value
  File "/Users/matthews/.pyenv/versions/pl/lib/python3.9/site-packages/h5py/_hl/group.py", line 483, in __setitem__
    ds = self.create_dataset(None, data=obj)
  File "/Users/matthews/.pyenv/versions/pl/lib/python3.9/site-packages/h5py/_hl/group.py", line 183, in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
  File "/Users/matthews/.pyenv/versions/pl/lib/python3.9/site-packages/h5py/_hl/dataset.py", line 48, in make_new_dset
    data = base.array_for_new_object(data, specified_dtype=dtype)
  File "/Users/matthews/.pyenv/versions/pl/lib/python3.9/site-packages/h5py/_hl/base.py", line 118, in array_for_new_object
    data = np.asarray(data, order="C", dtype=as_dtype)
  File "/Users/matthews/.pyenv/versions/pl/lib/python3.9/site-packages/torch/_tensor.py", line 970, in __array__
    return self.numpy()
```
</details>

I'm not sure exactly what we want to do, but I'm a fan of failing fast. If our plan is to only support autograd (pennylane.numpy) and vanilla numpy (which I think is our plan), I would just do something like this:
```python
if self.info["array_interface"] == "autograd":
    self.info["requires_grad"] = value.requires_grad
elif self.info["array_interface"] != "numpy":
    raise TypeError(f"expected array interface to be in ('numpy', 'autograd'), got {self.info['array_interface']} instead")
```

Reply:
Good call - limited support to autograd and numpy